src/Sparrow/Json/JsonContextPoolBase.cs

```diff
@@ -1,240 +1,34 @@
-            {
-                return _parents.Add(parent);
-            }
-
-            public void RemoveContextPool(JsonContextPoolBase<T> parent)
-            {
-                _parents.Remove(parent);
-            }
-
-            public void RemoveThreadIdHolder(ThreadIdHolder holder)
-            {
-                ThreadIdHolders.Remove(holder);
-            }
-        }
-
-        [ThreadStatic]
-        private static ContextStackThreadReleaser _releaser;
-        
-
-        private void EnsureCurrentThreadContextWillBeReleased(int currentThreadId)
-        {            
-            if (_releaser == null)
-            {
-                _releaser = new ContextStackThreadReleaser();
-            }
-
-            if (_releaser.AddContextPool(this) == false)
-                return;
-
-            while (true)
-            {
-                var copy = _threadIds;
-                for (int i = 0; i < copy.Length; i++)
-                {
-                    if (copy[i].ThreadId == -1)
-                    {
-                        if(Interlocked.CompareExchange(ref copy[i].ThreadId, currentThreadId, -1) == -1)
-                        {
-                            _releaser.ThreadIdHolders.Add(copy[i]);
-                            return;
-                        }
-                    }
-                }
-                
-                var threads = new ThreadIdHolder[copy.Length + 1];
-                Array.Copy(copy, threads, copy.Length);
-                threads[copy.Length] = new ThreadIdHolder
-                {
-                    ThreadId = currentThreadId
-                };
-
-                if (Interlocked.CompareExchange(ref _threadIds, threads, copy) == copy)
-                {
-                    _releaser.ThreadIdHolders.Add(threads[copy.Length]);
-                    break;
-                }
-            }
-        }
-
-        private class ContextStack : StackHeader<T>, IDisposable
-        {
-            public bool AvoidWorkStealing;        
-            public void Dispose()
-            {
-                GC.SuppressFinalize(this);
-                DisposeOfContexts();
-                // explicitly don't want to do this in the finalizer, if the instance leaked, so be it
-                _contextStackPool.Free(this);
-            }
-
-            private void DisposeOfContexts()
-            {
-                var current = Interlocked.Exchange(ref Head, HeaderDisposed);
-
-                while (current != null)
-                {
-                    var ctx = current.Value;
-                    current = current.Next;
-                    if (ctx == null)
-                        continue;
-                    if (!ctx.InUse.Raise())
-                        continue;
-                    ctx.Dispose();
-                }
-            }
-        }
+        private readonly T[][] _perCoreContexts;
+        private readonly ConcurrentQueue<T> _globalQueue = new ConcurrentQueue<T>();
```

Comment:
We have up to 64 slots per core, and then we fail over to a global queue. 
The idea is that we have an intentionally small per core value and then a shared system.
There is no work stealing here because we assume that the cores are going to be scheduled by the OS significantly.